analytics.py

```diff
@@ -10,52 +11,87 @@
 from .util.datetime_helpers import utc_now
 
 
-class Analytics(object):
+class Analytics:
+    """Loads configuration and dispatches methods for analytics providers.
+
+    SINGLETON!! Only one instance is meant to exist at any given time.
+
+    Configuration is loaded only on the first instantiation or when
+    `refresh=True` is passed in to facilitate reload.
+    """
+
+    _singleton_instance = None
 
     GLOBAL_ENABLED = None
-    LIBRARY_ENABLED = set()
-
-    def __init__(self, _db):
-        self.sitewide_providers = []
-        self.library_providers = defaultdict(list)
-        self.initialization_exceptions = {}
-        Analytics.GLOBAL_ENABLED = False
-        Analytics.LIBRARY_ENABLED = set()
+    LIBRARY_ENABLED: Set[int] = set()
+
+    def __new__(cls, _db, refresh=False) -> "Analytics":
+        instance = cls._singleton_instance
+        if instance is None:
+            refresh = True
+            instance = super().__new__(cls)
+            cls._singleton_instance = instance
+        if refresh:
+            instance._initialize_instance(_db)
+        return instance
+
+    @classmethod
+    def _reset_singleton_instance(cls):
+        """Reset the singleton instance. Primarily used for tests."""
+        cls._singleton_instance = None
+
+    def _initialize_instance(self, _db):
```

Comment:
It would be helpful to add some logging here so we could see in the logs when it's created and how much time it took to initialize.

Reply:
I added some logging and added a new decorator at `.util.log.log_elapsed_time`.